File: Construct3/Construct3/Library/Telerik.OpenAccess.SDK/IDEIntegrations/templates/templateObjectContainerProvider.vb
'Copyright (c) %CompanyName%.  All rights reserved.
'

Imports %ProductNamespace%
Imports %ProductNamespace%.Util

public class %ClassName%
	implements IObjectScopeProvider
	implements IObjectContainerProvider
	implements IDBConnection

	private _myDatabase as Database
	private _myScope as IObjectScope
	private myContainer as ObjectContainer
	private shared the%ClassName% as %ClassName%

	Public Sub New ()
    End Sub

    Public Sub New(ByVal aContainer As ObjectContainer)
        myContainer = aContainer
    End Sub

	' Allows a dynamically loaded assembly to open the database.
	' Must be called before the ObjectScope is aquired.
    Public Shared Sub AdjustForDynamicLoad()
        If the%ClassName% Is Nothing Then
            the%ClassName% = New %ClassName%()
        End If

        If the%ClassName%._myDatabase Is Nothing Then
            Dim assumedInitialConfiguration As String
            assumedInitialConfiguration = "<%RootNodeName%>" _
                           & "<references>" _
                           & "<reference assemblyname='PLACEHOLDER' configrequired='True'/>" _
                           & "</references>" _
                           & "</%RootNodeName%>"
            Dim repl As String = "PLACEHOLDER"
            Dim dll As System.Reflection.Assembly = the%ClassName%.GetType().Assembly
            assumedInitialConfiguration = assumedInitialConfiguration.Replace(repl, dll.GetName().Name)
            Dim xmlDoc As New System.Xml.XmlDocument
            xmlDoc.LoadXml(assumedInitialConfiguration)
			Dim allDlls As System.Reflection.Assembly() = New System.Reflection.Assembly() { dll }

            the%ClassName%._myDatabase = %ProductNamespace%.Database.Get("%ConnectionName%", _
											xmlDoc.DocumentElement, allDlls)
        End If
    End Sub

	public shared function Database() as Database 
		if the%ClassName% is nothing then
			the%ClassName% = new %ClassName%()
        end if

		if the%ClassName%._myDatabase is nothing then
			the%ClassName%._myDatabase = %ProductNamespace%.Database.Get( "%ConnectionName%" )
'		try
'			the%ClassName%.myDatabase.Properties.ConnectionTimeout = 10000
'		catch (Exception)
'
        end if

		return the%ClassName%._myDatabase
	end function

	public shared function ObjectScope() as IObjectScope 
		Database()

		if the%ClassName%._myScope is nothing then
			the%ClassName%._myScope = GetNewObjectScope()
        end if

		return the%ClassName%._myScope
	end function

	public shared function GetNewObjectScope() as IObjectScope 
		dim db as Database = Database()

		dim newScope as IObjectScope = db.GetObjectScope(%GetObjectScopeParameters%)
		return newScope
	end function

    Public Shared Function ObjectContainer() As ObjectContainer
        If the%ClassName% Is Nothing Then
            the%ClassName% = New %ClassName%
        End If
        If the%ClassName%.myContainer Is Nothing Then
            the%ClassName%.myContainer = GetNewObjectContainer()
        End If

        Return the%ClassName%.myContainer
    End Function

    Public Shared Function GetNewObjectContainer() As ObjectContainer
        Dim oc As ObjectContainer = New ObjectContainer
        oc.DBConnection = New %ClassName%(oc)
        oc.AutoSync = True
        Return oc
    End Function

#Region " IDBConnection Members "

    Public Function Load(ByVal data As FillInstructionData) As %ProductNamespace%.ObjectContainer.ChangeSet Implements IDBConnection.Load
		Dim os As IObjectScope	= %ClassName%.GetNewObjectScope()
        Dim ccData As %ProductNamespace%.ObjectContainer.ChangeSet

		os.Transaction.Begin()

'		try
			Dim fillInstr As FillInstruction = FillInstruction.Deserialize( data )

			ccData = fillInstr.Execute( os )
'		finally
			os.Transaction.Rollback()
			os.Dispose()

        Return ccData
    End Function

    Public Sub Save(ByVal data As %ProductNamespace%.ObjectContainer.ChangeSet) Implements IDBConnection.Save
		Dim os As IObjectScope	= %ClassName%.GetNewObjectScope()
		%ProductNamespace%.ObjectContainer.CommitChanges( data, %ProductNamespace%.ObjectContainer.Verify.Changed, os, true, false )
		os.Dispose()
    End Sub

    Public Function Sync(ByVal data As %ProductNamespace%.ObjectContainer.ChangeSet) As %ProductNamespace%.ObjectContainer.ChangeSet Implements IDBConnection.Sync
		Dim os As IObjectScope	= %ClassName%.GetNewObjectScope()
        Dim ccData As ObjectContainer.ChangeSet

		ccData = %ProductNamespace%.ObjectContainer.CommitChanges( data, %ProductNamespace%.ObjectContainer.Verify.Changed, os, true, true )

		os.Dispose()
        Return ccData
    End Function
#End Region
end class

File: Construct3/Construct3/Library/Telerik.OpenAccess.SDK/IDEIntegrations/templates/templateObjectContainerProvider.cs
//Copyright (c) %CompanyName%.  All rights reserved.
//
using System;
using System.Collections;
using %ProductNamespace%;
using %ProductNamespace%.Util;

namespace %DefaultNamespace%
{
	/// <summary>
	/// Summary description for %ClassName%.
	/// </summary>
	public class %ClassName% : IObjectScopeProvider, IObjectContainerProvider, IDBConnection
	{
		private Database myDatabase = null;
		private IObjectScope myScope = null;
		private ObjectContainer myContainer = null;

		static private %ClassName% the%ClassName% = null;

		public %ClassName%()
		{
		}

		public %ClassName%( ObjectContainer aContainer )
		{
			this.myContainer = aContainer;
		}

		static public Database Database()
		{
			if( the%ClassName% == null )
				the%ClassName% = new %ClassName%();

			if( the%ClassName%.myDatabase == null )
			{
				the%ClassName%.myDatabase = %ProductNamespace%.Database.Get( "%ConnectionName%" );
				try
				{ the%ClassName%.myDatabase.Properties.ConnectionTimeout = 10000;}
				catch (Exception)
				{}
			}
			return the%ClassName%.myDatabase;
		}

		static public IObjectScope ObjectScope()
		{
			Database();

			if( the%ClassName%.myScope == null )
				the%ClassName%.myScope = GetNewObjectScope();

			return the%ClassName%.myScope;
		}

		static public IObjectScope GetNewObjectScope()
		{
			Database db = Database();

			IObjectScope newScope = db.GetObjectScope(%GetObjectScopeParameters%);
			return newScope;
		}
		static public ObjectContainer ObjectContainer()
		{
			if( the%ClassName% == null )
				the%ClassName% = new %ClassName%();
			if( the%ClassName%.myContainer == null )
				the%ClassName%.myContainer = GetNewObjectContainer();

			return the%ClassName%.myContainer;
		}
		static public ObjectContainer	GetNewObjectContainer()
		{
			ObjectContainer	oc = new ObjectContainer();
			oc.DBConnection = new %ClassName%( oc );
			oc.AutoSync = true;
			return oc;
		}

        /// <summary>
		/// Adjusts for dynamic loading when no entry assembly is available/configurable.
		/// </summary>
		/// <remarks>
        /// When dynamic loading is used, the configuration path from the
        /// applications entry assembly to the connection setting might be broken.
        /// This method makes up the necessary configuration entries.
        /// </remarks>
        static public void AdjustForDynamicLoad()
        {
            if( the%ClassName% == null )
                the%ClassName% = new %ClassName%();

            if( the%ClassName%.myDatabase == null )
            {
                string assumedInitialConfiguration =
                           "<%RootNodeName%>" +
                               "<references>" +
                                   "<reference assemblyname='PLACEHOLDER' configrequired='True'/>" +
                               "</references>" +
                           "</%RootNodeName%>";
                System.Reflection.Assembly dll = the%ClassName%.GetType().Assembly;
                assumedInitialConfiguration = assumedInitialConfiguration.Replace(
                                                    "PLACEHOLDER", dll.GetName().Name);
                System.Xml.XmlDocument xmlDoc = new System.Xml.XmlDocument();
                xmlDoc.LoadXml(assumedInitialConfiguration);
                Database db = %ProductNamespace%.Database.Get("%ConnectionName%", 
                                            xmlDoc.DocumentElement,
                                            new System.Reflection.Assembly[] { dll } );

                the%ClassName%.myDatabase = db;
				try
				{ the%ClassName%.myDatabase.Properties.ConnectionTimeout = 10000;}
				catch (Exception)
				{}
            }
        }

		#region IDBConnection Members

		public %ProductNamespace%.ObjectContainer.ChangeSet Load( FillInstructionData instrData )
		{
			ObjectContainer.ChangeSet	ccData = null;
			IObjectScope				os = %ClassName%.GetNewObjectScope();

			os.Transaction.Begin();
			try
			{
				FillInstruction	fillInstr = FillInstruction.Deserialize( instrData );

				ccData = fillInstr.Execute( os );
			}
			finally
			{
				os.Transaction.Rollback();
				os.Dispose();
			}
			return ccData;
		}

		public void						Save( ObjectContainer.ChangeSet changes )
		{
			IObjectScope		os = %ClassName%.GetNewObjectScope();
			try
			{
				%ProductNamespace%.ObjectContainer.CommitChanges( changes, %ProductNamespace%.ObjectContainer.Verify.Changed, os, true, false );
			}
			finally
			{
				os.Dispose();
			}
		}

		public ObjectContainer.ChangeSet	Sync( ObjectContainer.ChangeSet changes )
		{
			ObjectContainer.ChangeSet	ccData = null;
			IObjectScope		os = %ClassName%.GetNewObjectScope();
			try
			{
				ccData = %ProductNamespace%.ObjectContainer.CommitChanges( changes, %ProductNamespace%.ObjectContainer.Verify.Changed, os, true, true );
			}
			finally
			{
				os.Dispose();
			}
			return ccData;
		}
		#endregion
	}
}
